### PR TITLE
Add measure checks

### LINKED ISSUE #340: Missing check for 1d and 2d triangulations
_From [bange...@gmail.com](https://code.google.com/u/111126369971033778052/) on November 26, 2014 18:32:44_

(Reported by Lukas Korous in an earlier pull request.)

In the 3d specialization of Triangulation::create_triangulation() (currently around line 1940 of tria.cc, we have this piece of code:

        // check that all cells have
        // positive volume. if not call the
        // invert_all_cells_of_negative_grid
        // and reorder_cells function of
        // GridReordering before creating
        // the triangulation
        for (unsigned int cell_no=0; cell_no<cells.size(); ++cell_no)
          AssertThrow (dealii::GridTools::cell_measure(triangulation.vertices,
                                                       cells[cell_no].vertices) >= 0,
                       ExcGridHasInvalidCell(cell_no));

First, the comment no longer describes what the piece of code does. Second, this piece of code is missing from the 1d and 2d specializations of the function, although it is clearly useful (at least if dim==spacedim) and should be added there.

_Original issue: http://code.google.com/p/dealii/issues/detail?id=266_

### PR DESCRIPTION
**Pull request by drwells**
This commit adds checks to a triangulation to verify that the cells have positive measure (as long as we do not keep track of distorted cells).

Perhaps unsurprisingly, adding these checks in found some problems with one of the mesh generators. I fixed those too.

Closes #340.

### CODE REVIEW

File: source/grid/grid_generator.cc [hunk 1]
```diff
@@ -856,26 +856,44 @@ namespace GridGenerator
...
+    // recompute both of them.
+    std_cxx11::array<Tensor<1, spacedim>, dim> compute_edges = edges;
+    std::vector<unsigned int> compute_subdivisions = subdivisions;
+    if (compute_subdivisions.size() == 0)
       {
-        std::vector<unsigned int> new_subdivisions(dim, 1);
-        subdivided_parallelepiped<dim,spacedim>(tria, origin, edges, new_subdivisions, colorize);
-        return;
+        compute_subdivisions.resize(dim);
+        std::fill(compute_subdivisions.begin(), compute_subdivisions.end(), 1);
```
Comment (kronbichler) [inline source/grid/grid_generator.cc]: Can't you simply write `compute_subdivisions.resize(dim,1)`, then you don't need to `fill` afterwards?

File: source/grid/grid_tools.cc [hunk 2]
```diff
@@ -167,6 +167,53 @@ namespace GridTools
...
+  {
+    Assert(false, ExcNotImplemented());
+    return std::numeric_limits<double>::quiet_NaN();
+  }
+
+
+
   template <>
   double
   cell_measure<3>(const std::vector<Point<3> > &all_vertices,
```
Comment (kronbichler) [inline source/grid/grid_tools.cc]: Why didn't you put the second template argument `3` here as well? You did in the 1D and 2D. ;-)

**Comment by kronbichler** [comment]
Apart from those minor comments, this looks good to me.

File: source/grid/grid_tools.cc (hunk 2 above)
Comment (drwells) [inline source/grid/grid_tools.cc]: That function was already there. I will make it consistent with the others (that change does make the code prettier).

File: source/grid/grid_generator.cc (hunk 1 above)
Comment (drwells) [inline source/grid/grid_generator.cc]: Good point (since it must be empty). I fixed this.

**Comment by kronbichler** [comment]
Looks good. Thanks for fixing this one!

File: include/deal.II/grid/grid_tools.h [hunk 3]
```diff
@@ -158,8 +158,8 @@ namespace GridTools
    * function also works for cells that do not exist except that you make it
    * up by naming its vertices from the list.
    */
-  template <int dim>
-  double cell_measure (const std::vector<Point<dim> > &all_vertices,
+  template <int dim, int spacedim=dim>
```
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h]: @drwells I see a problem here without C++11: http://cdash.kyomu.43-1.org/viewBuildError.php?buildid=608
We cannot have default template arguments on functions. As far as I can tell, there should be no problems to just omit the default value here. Could you please fix this in a follow-up PR?

Comment (drwells) [inline include/deal.II/grid/grid_tools.h]: Whoops. Today I learned something new about C++11.

**Pull request merged**